File: problems/binary-tree-path/binary_tree_path.c

#include <stdio.h>
#include <stdlib.h>

typedef struct TreeNode {
  int val;
  struct TreeNode *left;
  struct TreeNode *right;
} TreeNode;

TreeNode *create_node(int val) {
  TreeNode *node = (TreeNode *)malloc(sizeof(TreeNode));
  node->val = val;
  node->left = NULL;
  node->right = NULL;
  return node;
}

TreeNode *build_tree(int depth) {
  if (depth == 0)
    return NULL;
  TreeNode *node = create_node(depth);
  node->left = build_tree(depth - 1);
  node->right = build_tree(depth - 1);
  return node;
}

void free_tree(TreeNode *root) {
  if (!root)
    return;
  free_tree(root->left);
  free_tree(root->right);
  free(root);
}

int dfs(TreeNode *root, int sum) {
  if (!root)
    return 0;

  int current = sum * 10 + root->val;

  if (!root->left && !root->right) {
    return current;
  }

  return dfs(root->left, current) + dfs(root->right, current);
}

int main() {
  TreeNode *root = build_tree(12);
  int total = 0;

  for (int i = 0; i < 600; i++) {
    total += dfs(root, 0);
  }

  free_tree(root);
  printf("Total: %d\n", total);
  return 0;
}

File: problems/binary-tree-path/binary_tree_path.rs

struct TreeNode {
    val: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(val: i32) -> Self {
        TreeNode { val, left: None, right: None }
    }
}

fn build_tree(depth: i32) -> Option<Box<TreeNode>> {
    if depth == 0 {
        return None;
    }
    let mut node = TreeNode::new(depth);
    node.left = build_tree(depth - 1);
    node.right = build_tree(depth - 1);
    Some(Box::new(node))
}

fn dfs(root: &Option<Box<TreeNode>>, sum: i32) -> i32 {
    let node = match root {
        Some(n) => n,
        None => return 0,
    };
    
    let current = sum * 10 + node.val;
    
    if node.left.is_none() && node.right.is_none() {
        return current;
    }
    
    dfs(&node.left, current) + dfs(&node.right, current)
}

fn main() {
    let root = build_tree(12);
    let mut total = 0;
    for _ in 0..600 {
        total += dfs(&root, 0);
    }
    println!("Total: {}", total);
    // std::hint::black_box(total);
}
